core/src/server/rpc_module.rs

```diff
@@ -987,6 +1120,7 @@ impl SubscriptionSink {
 	///     let mut sink = pending.accept().unwrap();
 	///     let stream = futures_util::stream::iter(vec![1_usize, 2, 3]);
 	///     tokio::spawn(async move { sink.pipe_from_stream(stream).await; });
+	///     Ok(())
 	/// });
 	/// ```
 	pub async fn pipe_from_stream<S, T>(&mut self, stream: S) -> SubscriptionClosed
```

Comment:
Same here; does it make sense to use PipeFromStreamResult? it feels odd having two very similar `pipe_from_stream`/`pipe_from_try_stream` calls that result in different but similar return types, but maybe there's good reason for it?

These calls do take `&mut self` at the mo, and the versions on `PendingSink` consume self. I can imagine this being tricky.

Is there a way to make these APIs consistent? Should these take `mut self` too? @niklasad1 what do you reckon?

Reply:
yeah, that one is tricky for sure.

Recall that `PendingSubscription` must be moved to be converted to a `SubscriptionSink` so it's not possible to unify that API.

So maybe we could make the APIs: `Subscription::accept, Subscription::pipe_from_stream, Subscription::pipe_from_try_stream` private and folks you be forced the use `Pending::pipe_from_stream` at least then they won't be confused.

I don't see a reason why anyone would need those APIs anyway.

Reply:
I like the idea of having `Subscription::pipe_from_stream` private.
One drawback atm is that `Pending::pipe_from_stream` could not be reused, due to it taking `self` and not `&mut self`. Tho, I believe this limitation could be lifted to ensure the API is as permissive as `Subscription::pipe_from_stream`

Reply:
Going down this line, I'm wondering now whether it even makes sense to have a distinction between the `PendingSubscriptionSink` and `SubscriptionSink`.

How about we just return a `SubscriptionSink` straight away again, with the following:

- `pub fn accept(&mut self)` - accepts the connection and sends back the notification to the user without sending anything else. You call this if you want to send the accept message without sending anything else.
- the usual Sink impl would, if `accept` hasn't been explicitly called, call `accept` under the hood prior to sending any messages (if it hasn't been called already), so you don't need to explicitly `accept` at all if you don't care when the accept message is sent back.
- `pub fn pipe_from_stream`/`pipe_from_try_stream`  methods consume `&mut self`, and pretty much work as they do now, internally handling `accept()` too (if it hasn't been called already).
- `PipeFromStreamResult` no longer needs to contain the Sink anywhere. I'm hoping the lifetimes work out so that you can still do something like:

```rust
sink.pipe_from_stream(foo)
    .await
    .on_failure(|e| sink.close(e));
```

Or just trying to pipe a few things and then handling the result eventually (which I'd expect to basically come back the same whenever you try to pipe into a stream that's already been closed or whatever):

```rust
sink.pipe_from_stream(bar).await;
sink.pipe_from_stream(wibble).await;
sink.pipe_from_stream(foo).await
    .on_failure(|e| sink.close(e));
```

But possibly those `on_success`/`on_failure` calls needs a rethink they aren't very useful now.

What do you guys think?